packages/components/src/tree-grid/types.ts

```diff
@@ -0,0 +1,115 @@
+export type TreeGridRowProps = {
+	/**
+	 * The children to be rendered in the row.
+	 */
+	children: React.ReactNode;
+	/**
+	 * An integer value designating the level in the hierarchical tree structure.
+	 * Counting starts at 1. A value of `1` indicates the root level of the structure.
+	 */
+	level: React.AriaAttributes[ 'aria-level' ];
+	/**
+	 * An integer value that represents the position in the set.
+	 * A set is the count of elements at a specific level. Counting starts at 1.
+	 */
+	positionInSet: React.AriaAttributes[ 'aria-posinset' ];
+	/**
+	 * An integer value that represents the total number of items in the set,
+	 * at this specific level of the hierarchy.
+	 */
+	setSize: React.AriaAttributes[ 'aria-setsize' ];
+	/**
+	 * An optional value that designates whether a row is expanded or collapsed.
+	 * Currently this value only sets the correct aria-expanded property on a row,
+	 * it has no other built-in behavior.
+	 */
+	isExpanded?: React.AriaAttributes[ 'aria-expanded' ];
+};
+
+type RovingTabIndexItemPassThruProps = {
+	ref: React.ForwardedRef< any >;
+	tabIndex?: number;
+	onFocus: React.FocusEventHandler< any >;
+	[ key: string ]: any;
+};
+
+export type RovingTabIndexItemProps = {
+	/**
+	 * A render function that receives the props necessary to make it participate in the
+	 * roving tabindex. Any extra props will also be passed through to this function.
+	 *
+	 * Props passed as an argument to the render prop must be passed to the child
+	 * focusable component/element within the cell. If a component is used, it must
+	 * correctly handle the `onFocus`, `tabIndex`, and `ref` props, passing these to the
+	 * element it renders. These props are used to handle the roving tabindex functionality
+	 * of the tree grid.
+	 *
+	 * ```jsx
+	 * <TreeGridCell>
+	 * 	{ ( props ) => (
+	 * 		<Button onClick={ doSomething } { ...props }>
+	 * 			Do something
+	 * 		</Button>
+	 * 	) }
+	 * </TreeGridCell>
+	 * ```
+	 */
+	children?: ( props: RovingTabIndexItemPassThruProps ) => JSX.Element;
+	/**
+	 * If `children` is not a function, this component will be used instead.
+	 */
+	as?: React.ComponentType< RovingTabIndexItemPassThruProps >;
+	[ key: string ]: any;
```

Comment:
Honestly, in terms of TypeScript I think things could be a lot simpler if we didn't allow an `as` prop, and didn't allow pass-through of arbitrary props. All the same things would be possible with just a `children` render prop and no arbitrary pass-through props. We might want to revisit this API if these loose types cause more trouble than it's worth.

Reply:
Absolutely, let's keep this in mind